tests/torch/test_brevitas_qat.py

```diff
@@ -405,3 +492,102 @@ def test_brevitas_constant_folding(default_configuration):
             torch_inputset=data,
             configuration=default_configuration,
         )
+
+
+@pytest.mark.parametrize("manual_rounding", [None, 3])
+@pytest.mark.parametrize("power_of_two", [True, False])
+@pytest.mark.parametrize("n_bits", [4])
+@pytest.mark.parametrize("is_cnn", [True, False])
+def test_brevitas_power_of_two(
+    default_configuration,
+    manual_rounding: Optional[int],
+    power_of_two: bool,
+    n_bits: int,
+    is_cnn: bool,
+):
+    """Test a custom QAT network that uses power-of-two scaling.
+
+    Test whether a network using power-of-two scaling quantization is imported
+    correctly and roundPBS is used. Test that the Concrete ML does not override
+    the user's round PBS configuration.
+    """
+
+    net, x_all, _ = train_brevitas_network_tinymnist(is_cnn, n_bits, True, False, power_of_two)
+
+    # If rounding threshold is set -> nothing happens
+    # If Quantizer is not setup -> nothing happens
+    quantized_module = compile_brevitas_qat_model(
+        net.to("cpu"),
+        torch_inputset=x_all,
+        configuration=default_configuration,
+        rounding_threshold_bits=manual_rounding,
+    )
+
+    pot_should_be_applied = not manual_rounding and power_of_two
+    # Count the number of patterns that were optimized with roundPBS
+    num_round_pbs_layers = 0
+    for (_, node_op) in quantized_module.quant_layers_dict.values():
+        if isinstance(node_op, QuantizedMixingOp):
+            num_round_pbs_layers += 1 if node_op.rounding_threshold_bits is not None else 0
+            if pot_should_be_applied:
+                assert node_op.rounding_threshold_bits == node_op.lsbs_to_remove
+            elif manual_rounding:
+                # If manual rounding was set, LSBs_to_remove must be equal
+                # to the accumulator size minus the requested rounding_threshold_bits
+                assert node_op.rounding_threshold_bits == manual_rounding
+                assert node_op.lsbs_to_remove is not None
+
+    # Apply the PowerOfTwoScalingRoundPBSAdapter again. The second time
+    # the adapter will ignore already optimized patterns but report them
+    # as ignored.
+    adapter = PowerOfTwoScalingRoundPBSAdapter(quantized_module)
+    round_pbs_patterns = adapter.process()
+    # The power-of-two optimization will only work
+    # when Relu activations are used and scaling factors are forced to be 2**s
+    if pot_should_be_applied:
+        assert (
+            len(round_pbs_patterns) == 0
+        ), "Expected number of round PBS optimized patterns was not matched"
+        # 3 layers
+        assert (
+            adapter.num_ignored_valid_patterns == 3 - 1
+        ), "Expected number of ignored round PBS optimizable patterns was not matched"
+    else:
+        pass
+
+    # y_pred_clear_round = model.predict(x_test, fhe="disable")
```

Comment:
is the following supposed to be removed ?

Reply:
I added it back but correctness is not achieved, I'd rather push it as is and work on it later. I created an issue on it